src/engine/QueryPlanner.cpp

```diff
@@ -2459,6 +2473,46 @@ void QueryPlanner::GraphPatternPlanner::visitPathSearch(
   visitGroupOptionalOrMinus(std::move(candidatesOut));
 }
 
+// _______________________________________________________________
+void QueryPlanner::GraphPatternPlanner::visitSpatialSearch(
+    parsedQuery::SpatialQuery& spatialQuery) {
+  auto candidatesIn = planner_.optimize(&spatialQuery.childGraphPattern_);
+  std::vector<SubtreePlan> candidatesOut;
+
+  auto config = spatialQuery.toSpatialJoinConfiguration();
+
+  for (auto& sub : candidatesIn) {
+    // This helper function adds a subtree plan to the output candidates, which
+    // either has the child graph pattern as a right child or no child at all.
+    // If it has no child at all, the query planner may look for the right child
+    // of the SpatialJoin outside of the SERVICE. This is only allowed for max
+    // distance joins.
+    auto addCandidateSpatialJoin = [this, &sub, &config,
+                                    &candidatesOut](bool rightVarOutside) {
+      std::optional<std::shared_ptr<QueryExecutionTree>> right = std::nullopt;
+      if (!rightVarOutside) {
+        right = std::move(sub._qet);
+      }
+      auto spatialJoin = std::make_shared<SpatialJoin>(
+          qec_, std::make_shared<SpatialJoinConfiguration>(config),
+          std::nullopt, right);
+      auto plan = makeSubtreePlan<SpatialJoin>(std::move(spatialJoin));
+      candidatesOut.push_back(std::move(plan));
+    };
+
+    if (!spatialQuery.childGraphPattern_._graphPatterns.empty()) {
```

Comment:
As stated, I would prefer an explicit `optional` instead of the implicit  empty() (see my comments for the Magic Service thing.